### ISSUE TITLE
Poor performance in SnappyCodec because of string concat in tight loop

### ISSUE THREAD

**Issue by asfimport** [apache/parquet-java#1733]
String concatenation for SnappyUtil.validateBuffer which is called in a tight 
loop from both compress/decompress.

This shows up as taking 60% of time in the profiler 

```Java
  public static void validateBuffer(byte[] buffer, int off, int len) {
    Preconditions.checkNotNull(buffer, "buffer");
    Preconditions.checkArgument(off >= 0 && len >= 0 && off <= buffer.length - len,
        "Invalid offset or length. Out of buffer bounds. buffer.length=" + buffer.length
        + " off=" + off + " len=" + len);
  }
```

**Reporter**: [Cristian Opris](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=copris) / @cristianopris 
**Assignee**: [Ryan Blue](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rdblue) / @rdblue

<sub>**Note**: *This issue was originally created as [PARQUET-186](https://issues.apache.org/jira/browse/PARQUET-186). Please see the [migration documentation](https://issues.apache.org/jira/browse/PARQUET-2502) for further details.*</sub>

**Comment by asfimport** [comment]
[Julien Le Dem](https://issues.apache.org/jira/browse/PARQUET-186?#comment-14321095) / @julienledem:
Argh.
Preconditions.checkArgument should be replaced with an if to avoid generating the string when there is no error.

**Comment by asfimport** [comment]
[Ryan Blue](https://issues.apache.org/jira/browse/PARQUET-186?#comment-14321103) / @rdblue:
What about updating Preconditions to take an `Object...` argument and use `String.format`? That way we don't have to wrap all preconditions in an if for the same expression they check. See [`ValidationException.check`](https://github.com/kite-sdk/kite/blob/master/kite-data/kite-data-core/src/main/java/org/kitesdk/data/ValidationException.java#L49) for an example.

**Comment by asfimport** [comment]
[Julien Le Dem](https://issues.apache.org/jira/browse/PARQUET-186?#comment-14324794) / @julienledem:
This would still incur the cost of wrapping the parameters into an Array of Objects (ints into Integers and then array).
It could be negligible but we should verify this.

**Comment by asfimport** [comment]
[Ryan Blue](https://issues.apache.org/jira/browse/PARQUET-186?#comment-14347881) / @rdblue:
I just did a quick test for the overhead. I ran 10,000,000 iterations and observed the time using `System.nanoTime()` to record start and end times:

- With varargs, no args: 3868817 nanos
- With varargs, 2 args: 5954869 nanos
- Without varargs: 3745177 nanos

Re-running each seemed to produce approximately the same values. While it is slower, it isn't an order of magnitude slower and is significantly faster than string operations.
